bee-network/bee-autopeering/src/task.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    peer::{
+        lists::{ActivePeersList, ReplacementList},
+        PeerStore,
+    },
+    time::SECOND,
+};
+
+use priority_queue::PriorityQueue;
+use tokio::{sync::oneshot, task::JoinHandle, time};
+
+use std::{collections::HashMap, future::Future, time::Duration};
+
+pub(crate) const MAX_SHUTDOWN_PRIORITY: u8 = 255;
+const SHUTDOWN_TIMEOUT_SECS: Duration = Duration::from_secs(5 * SECOND);
+
+pub(crate) type ShutdownRx = oneshot::Receiver<()>;
+type ShutdownTx = oneshot::Sender<()>;
+
+pub(crate) type Repeat<T> = Box<dyn for<'a> Fn(&'a T) + Send>;
+
+/// Represents types driving an event loop.
+#[async_trait::async_trait]
+pub(crate) trait Runnable {
```

Comment:
A not for later but all of this will need to be replaced with backstage primitives soon

Reply:
I think autopeering will definitely become a plugin. It is designed to run autonomously, and simply needs a way to publish its results to the node, which then can act upon it (i.e. establish a gossip connection or something else depending on what the remote provides as a service). I don't know how "backstage" will be implemented, so maybe @pvdrz or @grtlr can give their input on how plugins and backstage work together.

Reply:
I see .... the autopeering plugin could of course also use "backstage" to implement its task business. Plugins in general are language independent eventually, so they can't rely on backstage (unless they are implemented in Rust, or do some FFI magic).

The question to me is now whether everything that spawns tasks should use actors/backstage. What's the scope of backstage actually? Should it be used everywhere now? I am a bit confused.

Reply:
Added the comment/todo.